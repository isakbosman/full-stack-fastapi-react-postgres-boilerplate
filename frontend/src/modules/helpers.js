// @flow
/**
 * Helper functions
 * @module Helpers
 */
import produce from 'immer';
import { format } from 'date-fns';

/**
 * Convert data attributes to Object
 * @param {Element} elem
 * @returns {{}}
 */
export function datasetToObject(elem: Element): Object {
  const data = {};
  [].forEach.call(elem.attributes, attr => {
    /* istanbul ignore else */
    if (/^data-/.test(attr.name)) {
      const camelCaseName = attr.name.substr(5).replace(/-(.)/g, ($0, $1) => $1.toUpperCase());
      data[camelCaseName] = attr.value;
    }
  });
  return data;
}

export function handleActions(actionsMap: Object, defaultState: Object): Function {
  return (state = defaultState, { type, ...rest }: Object = {}): Function =>
    produce(state, (draft): Object => {
      const action = actionsMap[type];
      let newState;

      if (action) {
        newState = action(draft, rest);
      }

      if (newState) {
        return newState;
      }

      return draft;
    });
}

export function keyMirror(obj: Object): Object {
  const output = {};

  for (const key in obj) {
    if (!Object.prototype.hasOwnProperty.call(output, key)) {
      output[key] = key;
    }
  }

  return output;
}

/**
 * Log grouped messages to the console
 * @param {string} type
 * @param {string} title
 * @param {*} data
 * @param {Object} [options]
 */
export function logger(type: string, title: string, data: any, options: Object = {}) {
  /* istanbul ignore else */
  if (process.env.NODE_ENV === 'development') {
    /* eslint-disable no-console */
    const { collapsed = true, hideTimestamp = false, typeColor = 'gray' } = options;
    const groupMethod = collapsed ? console.groupCollapsed : console.group;
    const parts = [`%c ${type}`];
    const styles = [`color: ${typeColor}; font-weight: lighter;`, 'color: inherit;'];

    if (!hideTimestamp) {
      styles.push('color: gray; font-weight: lighter;');
    }

    const time = format(new Date(), 'HH:mm:ss');

    parts.push(`%c${title}`);

    if (!hideTimestamp) {
      parts.push(`%c@ ${time}`);
    }

    /* istanbul ignore else */
    if (!window.SKIP_LOGGER) {
      groupMethod(parts.join(' '), ...styles);
      console.log(data);
      console.groupEnd();
    }
    /* eslint-enable */
  }
}

// $FlowFixMe
export const spread = produce(Object.assign);
