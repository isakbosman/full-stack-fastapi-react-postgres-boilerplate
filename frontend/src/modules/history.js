// @flow
import { createBrowserHistory } from 'history';
import qs from 'qs';

const history = createBrowserHistory();

history.location = {
  ...history.location,
  query: qs.parse(history.location.search.substr(1)),
  state: {},
};

/* istanbul ignore next */
history.listen(() => {
  history.location = {
    ...history.location,
    query: qs.parse(history.location.search.substr(1)),
    state: history.location.state || {},
  };
});

const { go, goBack, push, replace } = history;

export { go, goBack, push, replace };
export default history;
