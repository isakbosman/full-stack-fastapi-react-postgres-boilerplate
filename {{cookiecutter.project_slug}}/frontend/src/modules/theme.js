export const headerHeight = 70;

export const appColor = '#00b4d5';

export const easing = 'cubic-bezier(0.35, 0.01, 0.77, 0.34);';

export default {
  breakpoints: {
    xs: 0,
    ix: 400,
    md: 768,
    lg: 1024,
    xl: 1280,
    xxl: 1920,
  },
  palette: {
    primary: appColor,
  },
  button: {
    borderRadius: {
      xs: 4,
      lg: 28,
      xl: 32,
    },
    padding: {
      lg: [12, 28],
      xl: [14, 32],
    },
  },
};
