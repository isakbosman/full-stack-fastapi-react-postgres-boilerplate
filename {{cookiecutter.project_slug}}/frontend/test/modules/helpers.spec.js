import { advanceTo } from 'jest-date-mock';
import { datasetToObject, handleActions, keyMirror, logger, spread } from 'modules/helpers';

describe('helpers', () => {
  describe('datasetToObject', () => {
    const el = document.createElement('div');
    el.setAttribute('data-rule', 'yes');
    el.setAttribute('data-minute-maid', 'no');

    it('should convert DOMElement data to object', () => {
      expect(datasetToObject(el)).toMatchSnapshot();
    });
  });

  describe('handleActions', () => {
    it('should handle no changes', () => {
      const state = { isLogged: false };
      const reducer = handleActions(
        {
          LOGIN: draft => {
            draft.isLogged = true;
          },
        },
        state,
      );

      expect(reducer(state, { type: 'LOGOUT' })).toEqual({ isLogged: false });
    });

    it('should handle updates', () => {
      const state = { isLogged: false };
      const reducer = handleActions(
        {
          LOGIN: draft => {
            draft.isLogged = true;
          },
        },
        state,
      );

      expect(reducer(state, { type: 'LOGIN' })).toEqual({ isLogged: true });
    });

    it('should handle replace', () => {
      const state = { isLogged: false };
      const reducer = handleActions(
        {
          LOGIN: () => ({ test: true }),
        },
        state,
      );

      expect(reducer(state, { type: 'LOGIN' })).toEqual({ test: true });
    });
  });

  describe('keyMirror', () => {
    it('should return an object with mirrored keys', () => {
      expect(
        keyMirror({
          SWITCH_MENU: undefined,
          EXCEPTION: undefined,
          USER_LOGIN: undefined,
          USER_LOGIN_SUCCESS: undefined,
          USER_LOGIN_FAILURE: undefined,
          USER_LOGOUT: undefined,
        }),
      ).toMatchSnapshot();
    });
  });

  describe('logger', () => {
    let group;
    let groupCollapsed;
    let groupEnd;
    let log;

    beforeAll(() => {
      process.env.NODE_ENV = 'development';
      /* eslint-disable prefer-destructuring */
      group = console.group;
      groupCollapsed = console.groupCollapsed;
      groupEnd = console.groupEnd;
      log = console.log;
      /* eslint-enable */

      advanceTo(new Date(2019, 1, 1, 0, 0, 0));

      console.group = jest.fn();
      console.groupCollapsed = jest.fn();
      console.groupEnd = jest.fn();
      console.log = jest.fn();
    });

    afterAll(() => {
      process.env.NODE_ENV = 'test';
      console.group = group;
      console.groupCollapsed = groupCollapsed;
      console.groupEnd = groupEnd;
      console.log = log;
    });

    it('without options', () => {
      logger('type', 'Title', { a: 1 });

      expect(console.groupCollapsed).toHaveBeenLastCalledWith(
        '%c type %cTitle %c@ 00:00:00',
        'color: gray; font-weight: lighter;',
        'color: inherit;',
        'color: gray; font-weight: lighter;',
      );
      expect(console.log).toHaveBeenLastCalledWith({ a: 1 });
      expect(console.groupEnd).toHaveBeenCalledTimes(1);
    });

    it('with typeColor option', () => {
      logger('type', 'Title', ['a'], { typeColor: '#ffc52e' });

      expect(console.groupCollapsed).toHaveBeenLastCalledWith(
        '%c type %cTitle %c@ 00:00:00',
        'color: #ffc52e; font-weight: lighter;',
        'color: inherit;',
        'color: gray; font-weight: lighter;',
      );
      expect(console.log).toHaveBeenLastCalledWith(['a']);
      expect(console.groupEnd).toHaveBeenCalledTimes(2);
    });

    it('with hideTimestamp option', () => {
      logger('type', 'Title', 'OK', { hideTimestamp: true });

      expect(console.groupCollapsed).toHaveBeenLastCalledWith(
        '%c type %cTitle',
        'color: gray; font-weight: lighter;',
        'color: inherit;',
      );
      expect(console.log).toHaveBeenLastCalledWith('OK');
      expect(console.groupEnd).toHaveBeenCalledTimes(3);
    });

    it('with collapsed option', () => {
      logger('type', 'Title', null, { collapsed: false });

      expect(console.group).toHaveBeenLastCalledWith(
        '%c type %cTitle %c@ 00:00:00',
        'color: gray; font-weight: lighter;',
        'color: inherit;',
        'color: gray; font-weight: lighter;',
      );
      expect(console.log).toHaveBeenLastCalledWith(null);
      expect(console.groupEnd).toHaveBeenCalledTimes(4);
    });
  });

  describe('spread', () => {
    it('should merge properly', () => {
      const state = { dob: 1 };

      expect(spread(state, { dob: 1 })).toBe(state);

      expect(spread(state, { dob: 2 })).not.toBe(state);
    });
  });
});
