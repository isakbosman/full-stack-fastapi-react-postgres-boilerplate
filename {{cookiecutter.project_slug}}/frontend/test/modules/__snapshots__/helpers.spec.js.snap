// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers datasetToObject should convert DOMElement data to object 1`] = `
Object {
  "minuteMaid": "no",
  "rule": "yes",
}
`;

exports[`helpers keyMirror should return an object with mirrored keys 1`] = `
Object {
  "EXCEPTION": "EXCEPTION",
  "SWITCH_MENU": "SWITCH_MENU",
  "USER_LOGIN": "USER_LOGIN",
  "USER_LOGIN_FAILURE": "USER_LOGIN_FAILURE",
  "USER_LOGIN_SUCCESS": "USER_LOGIN_SUCCESS",
  "USER_LOGOUT": "USER_LOGOUT",
}
`;
